grouper/services/permission.py

```diff
@@ -29,6 +29,27 @@ def __init__(self, audit_log, permission_repository, permission_grant_repository
 
     def disable_permission(self, name, authorization):
         # type: (str, Authorization) -> None
+        """Disable a permission.
+
+        An invariant for permissions is that disabled permissions are not granted to anyone.  The
+        usecase is responsible for checking whether a permission is granted to any active groups or
+        service accounts before calling this service method, forcing the user disabling the
+        permission to remove those grants first.  However, grants to disabled groups are invisible
+        (by design) and retained only to be able to re-enable the group with its previous set of
+        permissions.  Semantically, it makes sense for those grants to quietly disappear without
+        explicit action.
+
+        Therefore, this service call first deletes all grants of the permission to disabled groups
+        before disabling the underlying permission.
+
+        All permission grants for a service account are deleted when that service account is
+        disabled, so this issue doesn't apply for service accounts.
+        """
+        grants = self.permission_grant_repository.revoke_inactive_group_grants_for_permission(name)
```

Comment:
I'm paranoid and would argue for explicitly checking for invariant violations (grants to enabled groups/service accounts) before this.

Reply:
Hm. I see why you'd want that, and I considered that, but then followed the rule that services will do anything that's possible in the database and invariant enforcement is the responsibility of the usecases. @cbguder, what do you think here? One could make an argument that one should think of this like a schema violation and enforce it in the repository.

Reply:
Yea had been noodling on this question before this PR. The model that makes the most sense to me is:
 - Usecases answer questions of _should_ something happen per the logic of the usecase.
 - Services answer questions of _can_ something happen, which means checking and enforcing **all** invariants.
 - Repositories just blindly do what they're told. If something would be a schema violation, it should be considered an invariant violation and therefore stopped at the service layer.

In particular, in order to stick to the principle of swappable backing repos, invariants _must_ be the same across repos. This leads to defining them at the service layer, making the service layer into a sort of thin waist of the stack.

With all that being said, I'm totally fine with using a different model. No matter which we choose though, I think we should settle on one soon and make it explicit, so we don't end up having a weird hodgepodge of checks like in the old Grouper code.

Reply:
It's because it would put Grouper in an inconsistent state: if a permission is granted to a disabled group, and then the permission is itself disabled, and then the group is re-enabled, what happens to the permission grant? Currently, we sprinkle `permission.enabled == True` checks all over the place, but the intended semantics of a permission is that re-enabling the permission doesn't restore any grants to groups. (In fact, we want to require that people remove the permission grant to all groups before disabling the permission.)

The root of this problem is that we preserve permission grants in disabled groups as a safety measure so that we can re-enable the group later and not lose all the permission grants, and do so by just special-casing deactivated groups. A better database model would probably move all of those permission grants to a separate table so they don't have to be excluded from every query, and only restore those grants when the group is re-enabled (or we just drop that feature entirely and add a huge warning to disabling the group). But both of those are major changes, way out of scope for this diff.

Reply:
Talked with @cbguder in person and he had a much better suggestion: change the service layer to unconditionally delete all grants, so it imposes the invariant without needing to do any checks. Then the usecase can do the check and prevent people from deleting a permission with active grants (business logic), and the service just makes it happen, including our consistency rule. I'll implement that.